### PR TITLE
Remove usage of legacy autograd function

### PR DESCRIPTION
**Pull request by yf225**
We are planning to put up a deprecation warning for legacy autograd function in 1.2: https://github.com/pytorch/pytorch/pull/22922. This PR removes all usage of legacy function in PyTorch core and test suite, to prepare for the eventual removal of legacy function.

### CODE REVIEW

**Comment by ezyang** [comment]
You should kill uses NestedIOFunction too, as it is also in legacy style.

**Comment by yf225** [comment]
I would like to scope out `NestedIOFunction` changes into another PR, as it's a user-visible class and needs more investigations in how to change it properly.

File: test/test_autograd.py [hunk 1]
```diff
@@ -504,31 +504,28 @@ def test_hook_none(self):
...
             was_called[0] = True
-        fn.register_hook(hook)
 
         x = torch.randn(5, 5, requires_grad=True)
         y = torch.randn(5, 5)
-        sum(fn(x, y)).sum().backward()
+        rx, ry = NoneGradientFunction.apply(x, y)
+        rx.register_hook(hook)
+        ry.register_hook(hook)
+        sum(rx, ry).sum().backward()
```
Comment (ezyang) [inline test/test_autograd.py]: Good!

File: test/test_autograd.py [hunk 2]
```diff
@@ -2098,29 +2098,32 @@ def assert_strict_equal(var1, var2):
...
 
         x = torch.randn(5, requires_grad=True)
-        a, b = F1()(x)
+        a, b = F1.apply(x)
         b = b + 1  # separate F1 from F2 by another op
         self.assertTrue(a.requires_grad)
         self.assertFalse(b.requires_grad)
-        c = F2()(a, b)
+        c = F2.apply(a, b)
+        print(c.grad_fn)
```
Comment (ezyang) [inline test/test_autograd.py]: You probably don't want this print

**Review by ezyang — APPROVED** [review]
Looks good, thank you. Note that this PR reduces our test coverage of legacy autograd functions, which means that we are less likely to notice breakages of legacy once we land it.

**Pull request closed without merging**